core/lib/zksync_core/src/consensus/testonly.rs

```diff
@@ -0,0 +1,291 @@
+            let block = storage
+                .fetch_block(ctx, validator::BlockNumber(i), self.operator_address)
+                .await?
+                .with_context(|| format!("missing block {i}"))?;
+            block.validate(validators, 1).unwrap();
+        }
+        Ok(())
+    }
+}
+
+async fn run_metadata_calculator(ctx: &ctx::Ctx, pool: ConnectionPool) -> anyhow::Result<()> {
+    const POLL_INTERVAL: time::Duration = time::Duration::milliseconds(100);
+    let mut n = L1BatchNumber(1);
+    while let Ok(()) = ctx.sleep(POLL_INTERVAL).await {
+        let mut storage = pool.access_storage().await.context("access_storage()")?;
+        let last = storage
+            .blocks_dal()
+            .get_sealed_l1_batch_number()
+            .await
+            .context("get_sealed_l1_batch_number()")?;
+
+        while n <= last {
+            let metadata = create_l1_batch_metadata(n.0);
+            storage
+                .blocks_dal()
+                .save_l1_batch_metadata(n, &metadata, H256::zero(), false)
+                .await
+                .context("save_l1_batch_metadata()")?;
+            n += 1;
+        }
+    }
+    Ok(())
+}
+
+impl StateKeeperRunner {
+    pub async fn run(self, ctx: &ctx::Ctx, pool: &ConnectionPool) -> anyhow::Result<()> {
+        scope::run!(ctx, |ctx, s| async {
+            let mut storage = pool.access_storage().await.context("access_storage()")?;
+            // ensure genesis
+            if storage
+                .blocks_dal()
+                .is_genesis_needed()
+                .await
+                .context("is_genesis_needed()")?
+            {
+                let mut params = GenesisParams::mock();
+                params.first_validator = self.operator_address;
+                ensure_genesis_state(&mut storage, L2ChainId::default(), &params)
+                    .await
+                    .context("ensure_genesis_state()")?;
+            }
+            let (stop_sender, stop_receiver) = sync::watch::channel(false);
+            let (miniblock_sealer, miniblock_sealer_handle) = MiniblockSealer::new(pool.clone(), 5);
+            let io = ExternalIO::new(
+                miniblock_sealer_handle,
+                pool.clone(),
+                self.actions_queue,
+                SyncState::new(),
+                Box::new(MockMainNodeClient),
+                self.operator_address,
+                u32::MAX,
+                L2ChainId::default(),
+            )
+            .await;
+            s.spawn_bg(miniblock_sealer.run());
+            s.spawn_bg(run_metadata_calculator(ctx, pool.clone()));
+            s.spawn_bg(
+                ZkSyncStateKeeper::without_sealer(
+                    stop_receiver,
+                    Box::new(io),
+                    Box::new(MockBatchExecutorBuilder),
+                )
+                .run(),
+            );
+            ctx.canceled().await;
+            let _ = stop_sender.send(true);
```

Comment:
Nit: You can use `send_replace` instead of `send` to signal that you allow for no receivers to be present.

Reply:
done